clients/src/main/java/org/apache/kafka/common/requests/AbstractRequest.java

```diff
@@ -93,7 +95,18 @@ public short version() {
     }
 
     public Send toSend(String destination, RequestHeader header) {
-        return new NetworkSend(destination, serialize(header));
+        // For UpdateMetadataRequest, the toSend method on the same object will be called many times, each time with a different destination
+        // value and a header containing a different correlation id.
+        ByteBuffer headerBuffer = serializeStruct(header.toStruct());
+        bobyBufferLock.lock();
+        try {
+            if (bodyBuffer == null) {
+                bodyBuffer = serializeStruct(toStruct()).array();
+            }
+        } finally {
+            bobyBufferLock.unlock();
+        }
+        return new NetworkSend(destination, new ByteBuffer[]{headerBuffer, ByteBuffer.wrap(bodyBuffer)});
```

Comment:
Since this optimization is only needed for UpdateMetadataRequest and maybe controller requests, should we move this logic to `UpdateMetadataRequest.java` or `AbstractControllerRequest.java`.

Reply:
That's a good point. I've moved the logic to the UpdateMetadataRequest.